dev_docs/kibana_platform_plugin_intro.mdx

```diff
@@ -0,0 +1,234 @@
+---
+id: kibPlatformIntro
+slug: /kibana-dev-docs/platform-intro
+title: Plugins and the Kibana platform
+summary: An introduction to the Kibana platform and how to use it to build a plugin.
+date: 2021-01-06
+tags: ['kibana','onboarding', 'dev', 'architecture']
+---
+
+From an end user perspective, Kibana is a tool for interacting with Elasticsearch, providing an easy way 
+to visualize and analyze data.
+
+From a developer perspective, Kibana is a platform that provides a set of tools to build not only the UI you see in Kibana today, but
+almost any UI that can be dreamt up. The platform provides developers the ability to build applications, or inject extra functionality into 
+already existing applications. Did you know that almost everything you see in the
+Kibana UI is built inside a plugin? If you removed all plugins from Kibana, you'd be left with an empty navigation menu, and a set of
+developer tools. The Kibana platform is a blank canvas, just waiting for a developer to come along and create something!
+
+![Kibana personas](assets/kibana_platform_plugin_end_user.png)
+
+## Plugins vs The Platform
+
+Kibana provides the most basic and fundmental tools neccessary for building a plugin with the "Core" platform. The Core platform is not a plugin itself.
+There is however, a lot of platform functionality provided by plugins that are not a part of Core.
+For example, the "data" plugin provides basic utilities to search, query, and filter data in Elasticsearch. This code is not part of Core, but is still very fundamental
+for many plugins, and we strongly encourage using this service over querying Elasticsearch directly.
+
+<DocCallOut title="So what goes in Core and what ends up in a plugin?">
+Our heuristic is that code _required_ to build a plugin goes in Core, and code that is _optional_ goes in a plugin.
+In reality, however, it's a grey area.
+</DocCallOut>
+
+The functionality that is exposed by plugins, for other plugins to use, is considered to be part of the Kibana platform!
```

Comment:
I don't think it's correct for us to consider all exports or plugin contracts to be part of the platform. There are solution-specific plugins that expose public interfaces that I don't think should be considered part of the platform. When discussing this as part of https://github.com/elastic/kibana/issues/71566 we only talked about a subset of plugins and core being moved to be under the `platform/` folder.

In my opinion, our code structure needs to be reorganized to properly reflect what we've discussed as being part of the platform because there isn't a clear way to differentiate what we consider to be a platform shared-service from the other solution-specific plugins.

Reply:
> I don't think it's correct for us to consider all exports or plugin contracts to be part of the platform.

I'm okay with brainstorming better messaging. How would you distinguish between platform code and the entirety of all "shared services"(?), before we have that folder structure that you linked to?

We could say it depends on which team owns the plugin but a) I don't think developers should have to know/care about our team structure (which changes frequently), and b) I don't think that will end up always being true.

Trying to define three types of shared code is more difficult than defining two, and requires more overhead to understand.  What is the benefit to a developer in distinguishing between "non-core platform code" and other shared services? Or "core platform" vs "non-core platform" code?
 
Our plugin structure makes a difference between core and all other shared services by splitting them up into two parameters `(core, dependencies)`. Is it helpful for a developer to distinguish between the services available in the second parameter?

## Ideal future brainstorm

Taking a moment to ignore all technical complexities, and think of the ideal situation, I would want to have platform code as the first parameter (generic code that is relevant to every solution) and non-generic, solution/plugin specific shared services in the second.  Non-platform shared services can be disabled, while platform services can't.

The technical problem is x-pack platform code can't be part of the first param because it's built as a plugin, but let's pretend we can get around that.

Then let's think about the folder structure. There wouldn't need to be a `platform/core` folder, just `platform`.

Then the terminology we could use is:
 - platform services
 - plugin services
 - all shared services (platform + plugins)
 
 If we think this is a good long term direction, I could switch this terminology around to call "platform" core only core platform code, and call non-core platform code as "plugin services". Perhaps with a caveat that some plugin services are very "platform-y", and will over time, be migrated into the core system. That we use plugin services to test out potential platform services while their APIs are still under frequent iteration, so when we do move them, officially, into the platform, they are somewhat stable.

Reply:
Took a stab at giving a more accurate, though slightly more confusing, description of the eco system today.